File: LeetCode/208ImplementTrie/ImplementTrie.py
class TrieNode:
    def __init__(self):
        self.children = [None for i in range(26)]
        self.endOfWord = False

class Trie:
    # Space:O(ln) l=avgLengthOfWord, n=numOfWords
    def __init__(self):
        self.root = TrieNode()        

    def insert(self, word: str) -> None:
        # Time:O(n), Space:O(n), n=numOfCharactersInWord
        current = self.root
        for letter in word:
            idx = ord(letter) - 97
            if not current.children[idx]:
                current.children[idx] = TrieNode()
            current = current.children[idx]
        current.endOfWord = True

    def search(self, word: str) -> bool:
        # Time:O(n), Space:O(1)
        current = self.root
        for letter in word:
            idx = ord(letter) - 97
            if not current.children[idx]:
                return False
            current = current.children[idx]
        if current.endOfWord:
            return True
        return False

    def startsWith(self, prefix: str) -> bool:
        # Time:O(n), Space:O(1)
        current = self.root
        for letter in prefix:
            idx = ord(letter) - 97
            if not current.children[idx]:
                return False
            current = current.children[idx]
        return True

File: LeetCode/208ImplementTrie/ImplementTrie.cpp
struct TrieNode {
    vector<TrieNode*> children = vector<TrieNode*>(26);
    bool endOfWord = false;
};

class Trie {
public:
    // Space:O(ln) l=avgLengthOfWord, n=numOfWords
    TrieNode *root = new TrieNode();

    Trie() {
        
    }
    
    void insert(string word) {
        // Time:O(n), Space:O(n), n=numOfCharactersInWord
        TrieNode *tempNode, *current = root;
        int idx = 0;
        for (int i = 0; i < word.size(); i++) {
            idx = word[i] - 'a';
            if (!current->children[idx]) {
                tempNode = new TrieNode();
                current->children[idx] = tempNode;
            }
            current = current->children[idx];
        }
        current->endOfWord = true;
    }
    
    bool search(string word) {
        // Time:O(n), Space:O(1)
        TrieNode *current = root;
        int idx = 0;
        for (int i = 0; i < word.size(); i++) {
            idx = word[i] - 'a';
            if (!current->children[idx]) {
                return false;
            }
            current = current->children[idx];
        }
        return current->endOfWord;
    }
    
    bool startsWith(string prefix) {
        // Time:O(n), Space:O(1)
        TrieNode *current = root;
        int idx = 0;
        for (int i = 0; i < prefix.size(); i++) {
            idx = prefix[i] - 'a';
            if (!current->children[idx]) {
                return false;
            }
            current = current->children[idx];
        }
        return true;
    }
};
